할인 등록을 개선하라

eastshine-high opened the pull request:
* 할인에 기간 속성을 추가하였습니다.
* 정률 할인과 정액 할인 모두 구현되도록 하였습니다.

고민과 시행착오가 있어 PR이 조금 늦었네요.

피드백 부탁드려요.

모두 스터디 고생하셨습니다~

hannut91 committed:
Define Discount domain model

유비쿼터스 언어를 보고 할인 도메인 모델을 만들었습니다.
할인 도메인 모델은 할인이 할인 대상 상품을 정할 수 있습니다.

hannut91 committed:
Define DiscountRepository

할인 레포지토리를 생성했습니다.

hannut91 committed:
Register discount

할인 관련 기능들을 관리하는 서비스를 만들었습니다.
할인을 등록할 수 있습니다.

hannut91 committed:
Create health check controller

서버가 살아있는지 확인할 수 있도록 Hello!를 반환하는 HelloController를
만들었습니다.

hannut91 committed:
Clear querydsl directory before build

빌드하기 전에 이미 querydsl 폴더가 있다면 삭제하고 빌드하도록 했습니다.

hannut91 committed:
Remove version information

`./gradlew assemble`할 때 버전 정보로 jar파일이 만들어져서, 버전을
변경할 때 마다 Dockerfile을 수정해야 합니다.
그래서 버전정보를 삭제했습니다.

hannut91 committed:
Dockerize project

배포를 위해서 도커 파일을 만들었습니다.

hannut91 committed:
Run GitHub Actions on push and pull request

푸시하고 풀리퀘스트를 만들었을 때 GitHub Actions을 실행하도록 했습니다.

etff committed:
add swagger config

- api 문서화를 위한 도구로 Swagger를 의존성에 추가했습니다.

hannut91 committed:
Change heal check url to /health-check from root

API 루트는 다른 것을 보여줄 수 있는 경우가 있어서, 헬스 체크 url을 따로
사용하도록 수정했습니다.

hannut91 committed:
Rename to apply from attach

할인을 적용할 수 있다로 표현하기로 하여 apply로 변경합니다.

hannut91 committed:
Reveal ubiquitous language in test names

테스트 이름이 너무 기능적인 부분만 확인하는 것 같아서, 유비쿼터스 언어가
드러날 수 있도록 변경했습니다.

etff committed:
git keep 제거

JamieShin0201 committed:
Handle discount registration request

할인 등록 정보와 함께 POST /discounts 요청 시 할인을 등록합니다.

etff committed:
정책을 추가하라

- 정책에 정률할인과 정액할인을 추가한다

etff committed:
기간을 추가하라

etff committed:
쿠폰번호를 추가하라

정률을 비율을 추가하라

쿠폰 번호 테스트를 추가한다

비율의 동등성 수정과 인자를 수정하라

쿠폰의 수량을 추가하라

매직넘버를 제거한다

기간 빈 값 입력시 예외 처리

할인 값을 추가한다

etff committed:
쿠폰 도메인을 추가한다

쿠폰 이름을 추가한다

쿠폰 영속화기능을 추가한다

쿠폰 정책 기능을 추가하라

- 정률 정책을 추가한다.
- 정액 정책을 추가한다.

etff committed:
쿠폰 정책을 상속으로 변경하라

- 정률, 정액 정책으로 변경한다.

쿠폰 구분 컬럼을 수정하라

etff committed:
쿠폰 번호를 쿠폰 코드로 용어를 수정한다.

> https://github.com/CodeSoom/DDD-Kurly-Clone-Promotion/discussions/7#discussioncomment-1513859

etff committed:
정책 생성 테스트를 추가한다

etff committed:
동일성 비교 기능을 추가한다

etff committed:
쿠폰번호를 랜덤생성하지 않고 키워드로서 입력받도록 수정한다

etff committed:
유비쿼터스 언어를 반영하도록 용어를 수정한다

etff committed:
정책 타입의 중복 컬럼을 제거하라

etff committed:
오타수정

etff committed:
유효성 검사를 추가하라

etff committed:
불필요한 여백을 제거합니다

etff committed:
기준 유효성 중복검사를 제거하라

etff committed:
오타를 수정하라

etff committed:
최소 적용 금액을 추가한다

etff committed:
이름과 키워드의 중복검사를 추가하라

eastshine-high committed:
할인 기간을 추가하라

eastshine-high committed:
할인을 정액할인과 정률할인으로 세분화하라

할인을 상속구조로 변경하여 정률할인과 정액할인을 구현

eastshine-high committed:
세분화한 할인에 맞추어 할인 등록을 개선하라

팩토리를 만들어 정률할인과 정액할인을 생성하도록 변경.

eastshine-high committed:
MapStruct 의존성 추가

eastshine-high committed:
세분화한 할인에 맞추어 할인 등록 controller 로직 개선

할인 등록 서비스의변경된 시그니처에 맞게 호출 메서드 수정
Mapstruct를 적용하여 DTO를 Command에 매핑하도록 처리

eastshine-high committed:
충돌 해결

eastshine-high committed:
기본 생성자 제거

eastshine-high commented on src/main/java/com/kurly/promotion/domain/vo/Period.java:
```diff
@@ -0,0 +1,61 @@
+package com.kurly.promotion.domain.vo;
```
루트 애그리게잇과의 관계 파악을 쉽게 하기 위해서 패키지를 만들어봤습니다.

etff commented on src/main/java/com/kurly/promotion/domain/Discount.java:
```diff
@@ -1,34 +1,43 @@
+ */
 @Getter
-@AllArgsConstructor
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
+@AllArgsConstructor
+@DiscriminatorColumn
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
```
지난 PR때는 말씀을 못드렸는데, `@AllArgsConstructor` 를 썼을때 어떤점이 좋지않을까요?
혹시 설명해주실 수 있을까요?

eastshine-high replied:
@AllArgsConstructor이 위험한 이유는 클래스의 필드 변수 위치가 바뀔 경우, 이 생성자를 사용하는 측에서 순서가 바뀐 것을 인지하지 못할 수도 있는 점입니다.

etff commented on src/main/java/com/kurly/promotion/domain/DiscountCommand.java:
```diff
@@ -0,0 +1,56 @@
+ * Discount 도메인 객체의 생성을 정의
+ */
+public class DiscountCommand {
+
```
커맨드라는 용어는 모호합니다. 실제 주석과 용어정의가 다른 것같습니다.

etff replied:
커맨드 객체와 헷갈리기 때문에 힘든데요, CQRS 란 무엇인지 조사해보는 것도 좋을 것같습니다.

eastshine-high replied:
> 커맨드라는 용어는 모호합니다. 실제 주석과 용어정의가 다른 것같습니다.

맞습니다. ``Discount 도메인의 입력``을 책임집니다.

eastshine-high replied:
> 커맨드 객체와 헷갈리기 때문에 힘든데요, CQRS 란 무엇인지 조사해보는 것도 좋을 것같습니다.

Command는 CQRS가 맞습니다.
구현하기 전에 다른 DDD 프로젝트도 참고를 해보았는데요. 괜찮아 보이는 것을 사용했었는데. 이게 뭔지도 몰랐었네요ㅜㅜ

etff commented on src/main/java/com/kurly/promotion/domain/DiscountCommand.java:
```diff
@@ -0,0 +1,56 @@
+        private final LocalDate startDate;
+        private final LocalDate endDate;
+        private final Integer fixedAmout;
+        private final Integer flatRate;
```
팩터리 객체가 정적 클래스일 필요가 있을까요?

etff replied:
할인 등록은 할인의 책임으로만 두는 것이 어떨까요?

eastshine-high replied:
> 팩터리 객체가 정적 클래스일 필요가 있을까요?

클래스 안에 있는 클래스의 인스턴스화를 막기 위해서 static을 선언하였습니다.
클래스를 계층형으로 ``DiscountCommand.RegisterDiscount``, ``DiscountCommand.ModifyDiscount`` 표현하고자 했었습니다.

이렇게 하면 ``RegisterDiscountCommand`` ``ModifyDiscountCommand``와 같이 계속해서 클래스를 생성하지 않아도 되는 장점을 가질 수 있는 것 같습니다.

eastshine-high replied:
> 할인 등록은 할인의 책임으로만 두는 것이 어떨까요?

네 그것도 괜찮을 것 같습니다.
개인적인 생각은 할인 클래스에서는 할인 클래스의 생성자까지만 담당하고 다른 VO클래스와의 조합은 다른 곳에서 하면 어떨까 하는 의도였습니다.

etff commented on src/main/java/com/kurly/promotion/domain/DiscountCommand.java:
```diff
@@ -0,0 +1,56 @@
+
+    @ToString
+    @Getter
+    @Builder
```
특별한 의도가 있지않는이상 클래스위에 무조건적으로 롬복 어노테이션을 두는 것은 좋지 않습니다.

etff commented on src/main/java/com/kurly/promotion/domain/DiscountFactory.java:
```diff
@@ -0,0 +1,19 @@
+        switch(type){
+            case FIXED_AMOUNT: return requestDiscount.createFixedAmoutDiscount();
+            case FLAT_RATE : return requestDiscount.createFlatRateDiscount();
+            default: new IllegalArgumentException("할인 타입은 필수 파라미터입니다.");
+        }
+        return null;
```
null 을 반환하는것은 적절할지 생각해봅시다

eastshine-high replied:
> null 을 반환하는것은 적절할지 생각해봅시다

default 문장이 있어 이 문장은 수행하지 않는데요.
null보다 빈 Discount 객체를 반환하는 것이 괜찮을까요?
다른 방법은 떠오르지 않네요.

etff commented on src/main/java/com/kurly/promotion/domain/FixedAmountDiscount.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+@Entity
+@DiscriminatorValue("fixedAmount")
+public class FixedAmountDiscount extends Discount {
```
DiscriminatorValue 는 구분값을 두기위해 씁니다. 이렇게하면 디비에 snakeCase로 들어갑니다. DB에서 어떤 형태로 저장되는지 합의가 된것일까요?

eastshine-high replied:
> DiscriminatorValue 는 구분값을 두기위해 씁니다. 이렇게하면 디비에 snakeCase로 들어갑니다. DB에서 어떤 형태로 저장되는지 합의가 된것일까요?

병호님과는 조금은 다르게 해보자는 의식을 너무 강하게 했던 것 같네요.
DiscriminatorValue는 coupon 도메인과 같은 값으로 수정하였습니다.

etff commented on src/main/java/com/kurly/promotion/domain/FixedAmountDiscount.java:
```diff
@@ -0,0 +1,35 @@
+
+        if(fixedAmount == null){
+            throw new InvalidParameterException("빈 값을 입력할 수 없습니다.");
+        }
```
하드 코딩을 없애보면 어떨까요?
https://tecoble.techcourse.co.kr/post/2020-05-07-avoid-hard-coding/

eastshine-high replied:
> 하드 코딩을 없애보면 어떨까요?
> https://tecoble.techcourse.co.kr/post/2020-05-07-avoid-hard-coding/


이부분은 InvalidParameterException의 매개변수라는 생각으로 수정해 보았습니다.
```
throw new InvalidParameterException("FixedAmountDiscount.fixedAmount");
```

etff commented on src/main/java/com/kurly/promotion/domain/FlatRateDiscount.java:
```diff
@@ -0,0 +1,43 @@
+ */
+@Getter
+@DiscriminatorValue("Rate")
+@Entity
```
여기도

etff commented on src/main/java/com/kurly/promotion/domain/FlatRateDiscount.java:
```diff
@@ -0,0 +1,43 @@
+    @Column(name = "flat_rate", nullable = false)
+    private Integer flatRate;
+
+    @Builder
+    public FlatRateDiscount(Period period, Integer flatRate) {
```
builder는 어떤 용도로 쓰면 좋을까요?

eastshine-high replied:
> builder는 어떤 용도로 쓰면 좋을까요?

생성자에 매개변수가 많을 때 빌더를 고려합니다.
또 인스턴스를 초기화할 때 빌더를 사용하면 의미가 분명해지는 것 같습니다.

etff commented on src/main/java/com/kurly/promotion/domain/vo/Period.java:
```diff
@@ -0,0 +1,61 @@
+@ToString
+@Getter
+@Embeddable
+@NoArgsConstructor
+public class Period {
```
빈 생성자를 만드는 이유를 생각해보면 좋을 것같아요

eastshine-high replied:
> 빈 생성자를 만드는 이유를 생각해보면 좋을 것같아요

Dozzermapper같은 API를 사용할 때 빈생성자를 필요로 했었습니다.
또 저의 경우에는 테스트를 할 때 의미를 나타낼 필요가 없는 매개변수를 생성할 때 빈생성자를 사용했습니다.

etff commented on src/main/java/com/kurly/promotion/domain/vo/Period.java:
```diff
@@ -0,0 +1,61 @@
+
+/**
+ * 기간
+ */
```
javadoc 주석의 마지막은 `.`로 마무리해주세요

etff commented on src/main/java/com/kurly/promotion/ui/DiscountController.java:
```diff
@@ -33,9 +37,16 @@
+        DiscountCommand.RegisterDiscount registerDiscount = discountDtoMapper.of(discountRegistrationData);
+        Long productId = discountRegistrationData.getProductId();
+        DiscountType type = discountRegistrationData.getDiscountType();
+
```
실제 객체 맵핑은 쉽게 되지않습니다. 이렇게 복잡하게되면 굳이 mapper를 써야하나라는 생각도 들게됩니다.

eastshine-high replied:
네. 이 부분은 다른 방법도 좋아보입니다.

etff commented on src/test/java/com/kurly/promotion/application/DiscountServiceTest.java:
```diff
@@ -19,34 +26,47 @@
-
-        discountService = new DiscountService(discountRepository);
+        //this.discount = new Discount(1L, FLAT_RATE, PRODUCT_ID);
+        discountService = new DiscountService(new DiscountFactory(), discountRepository);
```
쓰지 않는 주석은 지워주세요

etff commented on src/test/java/com/kurly/promotion/controller/DiscountControllerTest.java:
```diff
@@ -0,0 +1,82 @@
+        given(discountService.registerDiscount(any(DiscountCommand.RegisterDiscount.class), any(Long.class), any(DiscountType.class)))
+                .willReturn(registeredId);
+        System.out.println("----------");
+        System.out.println(discountRegistrationData);
```
디버그가 필요하다면 디버깅 기능을 해보세요.
코드에 시스템 함수를 호출하지 않도록 하고 그 이유를 찾아보세요

eastshine-high replied:
> 디버그가 필요하다면 디버깅 기능을 해보세요.
> 코드에 시스템 함수를 호출하지 않도록 하고 그 이유를 찾아보세요

네 결국 디버깅으로 해결한 것들도 좀 있었네요.

etff commented on src/test/java/com/kurly/promotion/domain/FlatRateDiscountTest.java:
```diff
@@ -0,0 +1,51 @@
+    /*private static final Period PERIOD = Period.builder()
+            .startDate(LocalDate.of(2030, 1, 1))
+            .endDate(LocalDate.of(2030, 12, 1))
+            .build();*/
```
여기도 주석을 지워주세요

eastshine-high replied:
스터디가 마지막이라는 말에 성급히 커밋을 해버렸네요ㅜㅜ

etff commented on build.gradle:
```diff
@@ -38,6 +38,10 @@ dependencies {
+    annotationProcessor(
+            'org.projectlombok:lombok',
+            'org.projectlombok:lombok-mapstruct-binding:0.1.0'
+    )
```
추가한 이유를 알수있을까요?

eastshine-high replied:
https://mapstruct.org/documentation/stable/reference/html/#lombok
맵스트럭트를 이용할 때 롬복을 이용한 클래스가 있어 오류가 나지 않을까 해서 선언해 주었습니다.
그런데 ``annotationProcessor org.projectlombok:lombok``는 중복선언되어 있었네요..
그래서 중복된 부분은 제거하였습니다.

etff reviewed (changes requested):
고생하셨습니다. 생각해볼만 코멘트를 남겼습니다.

etff commented:
고생하셨습니다. 스터디가 종료되어 PR은 닫아도 될것같아요

eastshine-high committed:
중복선언된 DiscountController.java를 삭제하라

/controller/DiscountController.java(삭제)
/ui/DiscountController.java
두개가 선언되어 있었습니다.

eastshine-high committed:
주석을 제거하라 FlatRateDiscountTest.java

eastshine-high committed:
오타를 수정하라

eastshine-high committed:
하드코딩을 제거하라

eastshine-high committed:
주석에 . 추가하라

eastshine-high committed:
네이밍을 변경하라

eastshine-high committed:
변수 정리 및 주석을 제거하라

eastshine-high committed:
DiscountControllerTest의 오류를 개선하라

eastshine-high commented:
> 고생하셨습니다. 스터디가 종료되어 PR은 닫아도 될것같아요

감사합니다. 혹시 마무리를 지어봐도 괜찮을까요?

eastshine-high commented on src/test/java/com/kurly/promotion/ui/DiscountControllerTest.java:
```diff
@@ -98,8 +117,8 @@ void setUp() {
             void It_responses_discounts() throws Exception {
                 mockMvc.perform(get("/discounts"))
                         .andExpect(status().isOk())
-                        .andExpect(jsonPath("$[0].id").exists())
-                        .andExpect(jsonPath("$[0].flatRate").exists())
+                        //.andExpect(jsonPath("$[0].id").exists())
+                        //.andExpect(jsonPath("$[0].flatRate").exists())
                         .andExpect(jsonPath("$[0].productId").exists());
```
이 부분은 테스트 오류로 잠시 주석 처리를 하였는데요.
조회를 할 때 ``Discount``객체를 반환하면 하위 클래스``FlatRateDiscount`` 객체의 멤버 변수인 ``flatRate``는 접근을 할 수 없어서 생기는 오류인 것 같습니다.

이 문제는 어떻게 해결하는 것이 좋은 방법일까요?
JPA에서 상속 관계를 단일 테이블 전략으로 매핑했을 때, 조회를 어떻게 처리할 지를 잘 모르겠네요.

Discount 엔티티는 단일 테이블 전략으로 선언해 주었습니다.
```
@DiscriminatorColumn
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@Entity
public class Discount {
```

eastshine-high reviewed (commented):
피드백을 반영해 보앗습니다.
